src/main/host/syscall_types.rs

```diff
@@ -328,10 +328,16 @@ impl<T> TypedPluginPtr<T> {
 #[derive(Debug, PartialEq, Eq)]
 pub enum SyscallError {
     Errno(nix::errno::Errno),
-    Cond(SysCallCondition),
+    Blocked(Blocked),
     Native,
 }
```

Comment:
How does this handle the case in the C code where you have a `(SysCallReturn){.state = SYSCALL_DONE, .retval = -EINTR, .restartable = scr.restartable}` (equivalent to a `SyscallError::Errno` with a `restartable` value)? I think this might be less confusing with a new `SyscallError::Restart` variant instead of making a special case out of `-EINTR`.

Also, I think that converting between types will lose information. For example if you have a variable `ret` which is a `(SysCallReturn){.state = SYSCALL_DONE, .retval = -EINTR, .restartable = true}`, I don't think that `assert_eq!(ret, SyscallResult::from(ret).into())` would hold. There is the `test_syscallreturn_roundtrip` test in `descriptor/mod.rs`, but it doesn't test this case. There is some code such as [`log_syscall`](https://github.com/shadow/shadow/blob/49fb5c3e0f522b082dfca9fa71d583ca906bac23/src/main/host/syscall/format.rs#L572) which relies on this conversion not losing information (it takes an owned `c::SysCallReturn` and returns a round-trip owned `c::SysCallReturn` since the rust code has no way of working with a reference/ptr to a `c::SysCallReturn`, only an owned `c::SysCallReturn`).

Reply:
> How does this handle the case in the C code where you have a `(SysCallReturn){.state = SYSCALL_DONE, .retval = -EINTR, .restartable = scr.restartable}` (equivalent to a `SyscallError::Errno` with a `restartable` value)?

Hmm, this was somewhat intentional but I see the confusion and how this might lead to problems down the road. Right now handlers themselves never return `(SysCallReturn){.state = SYSCALL_DONE, .retval = -EINTR, .restartable = true}`; they only set the `restartable` flag on `SYSCALL_BLOCKED` results, and the `syscall_handler` code rewrites the `SYSCALL_BLOCKED` into `SYSCALL_DONE` if there's an unblocked syscall pending.

But yeah, this is made trickier because handlers *can* themselves check if unblocked signals are pending and return `-EINTR` themselves, and some need to (e.g. nanosleep). Currently the only ones that do this are also not restartable, but I'm not sure that'll always be the case.

Still working on this, but I'm leaning towards having the `restartable` flag in both the `Blocked` and `Failed` case:

```
 #[derive(Debug, PartialEq, Eq)]
 pub enum SyscallError {
-    Errno(nix::errno::Errno),
+    Failed(Failed),
     Blocked(Blocked),
     Native,
 }
...
+#[derive(Debug, PartialEq, Eq)]
+pub struct Failed {
+    pub errno: nix::errno::Errno,
+    pub restartable: bool,
+}
```

>  I think this might be less confusing with a new `SyscallError::Restart` variant instead of making a special case out of `-EINTR`.

We would still need a `restartable` flag in the `Blocked` case, unless we want to force every restartable signal handler that might otherwise return `Blocked` to also check whether there's an unblocked signal pending. It seems better to have that logic in one place.

> Also, I think that converting between types will lose information. For example if you have a variable `ret` which is a `(SysCallReturn){.state = SYSCALL_DONE, .retval = -EINTR, .restartable = true}`, I don't think that `assert_eq!(ret, SyscallResult::from(ret).into())` would hold. There is the `test_syscallreturn_roundtrip` test in `descriptor/mod.rs`, but it doesn't test this case. There is some code such as [`log_syscall`](https://github.com/shadow/shadow/blob/49fb5c3e0f522b082dfca9fa71d583ca906bac23/src/main/host/syscall/format.rs#L572) which relies on this conversion not losing information (it takes an owned `c::SysCallReturn` and returns a round-trip owned `c::SysCallReturn` since the rust code has no way of working with a reference/ptr to a `c::SysCallReturn`, only an owned `c::SysCallReturn`).

Ok, I'll be mroe careful about this. I think my current plan to add the `restartable` flag to both the `Blocked` and `Failed` cases will address it. I'll look at adding some more round trip tests as well...

Reply:
Finished making this change. I admit it's still ugly, but it's the least ugly option I can think of at the moment...

Reply:
Sounds good to me. I don't see a nicer way of handling this either.